### PR TITLE
rsmeter: support multiple datasinks

### PR DESCRIPTION
**Pull request by zhongzc**
Signed-off-by: Zhenchi <zhongzc_arch@outlook.com>

### What is changed and how it works?

ref #11691

What's Changed:

- `SingleTargetDataSink` as a worker. It registers/deregisters to the reporter according to the receiver address config.
- Add `Task::DataSinkReg` to `Reporter::Task` and the corresponding handling logic. The reporter registers/deregisters to the recorder according to the count of registered `DataSink`.
- Refer to `init_recorder`, extract helper construct functions `init_reporter` and `init_single_target`.

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/resource_metering/src/reporter/single_target.rs [hunk 1]
```diff
@@ -94,24 +116,93 @@ impl DataSink for SingleTargetDataSink {
...
+            return;
+        }
+
+        if self.address != new_address {
+            self.address = new_address;
+            let channel = {
+                let cb = ChannelBuilder::new(self.env.clone())
+                    .keepalive_time(Duration::from_secs(10))
+                    .keepalive_timeout(Duration::from_secs(3));
+                cb.connect(&self.address)
```
Comment (breeswish) [inline components/resource_metering/src/reporter/single_target.rs:133]: Previously this is called at TrySend, but now it is called when address is changed. Will it cause problems when connect takes long time?

File: components/resource_metering/src/reporter/single_target.rs [hunk 2]
```diff
@@ -94,24 +116,93 @@ impl DataSink for SingleTargetDataSink {
...
+            let data_sink = Box::new(DataSinkImpl {
+                scheduler: self.scheduler.clone(),
+            });
+            self.data_sink = Some(self.data_sink_reg.register(data_sink));
+        }
+    }
 
+    fn reset(&mut self) {
+        self.data_sink = None;
+        self.client = None;
```
Comment (breeswish) [inline components/resource_metering/src/reporter/single_target.rs:148]: I'm not familiar with this. Seems that here we simply drop it. Is it fine and will there be resource leaks?

**Comment by breeswish** [comment]
The rest LGTM

File: components/resource_metering/src/reporter/single_target.rs [hunk 3]
```diff
@@ -3,73 +3,95 @@
...
+
+    fn run(&mut self, task: Self::Task) {
+        match task {
+            Task::Records(records) => self.handle_records(records),
+            Task::ChangeAddress(address) => self.update_data_sink_and_client(address),
+        }
+    }
+
+    fn shutdown(&mut self) {
+        self.reset();
```
[the remark this answers is not in the document]
Comment (zhongzc) [reply components/resource_metering/src/reporter/single_target.rs:30]: It's ok. Nothing wrong will happen.

File: components/resource_metering/src/reporter/single_target.rs (hunk 2 above)
Comment (zhongzc) [reply components/resource_metering/src/reporter/single_target.rs:148]: It's fine. Dropping the inner channel will release resources.

File: components/resource_metering/src/reporter/single_target.rs (hunk 1 above)
Comment (zhongzc) [reply components/resource_metering/src/reporter/single_target.rs:133]: I will move it to `try_send`.

File: components/resource_metering/src/reporter/data_sink_reg.rs [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    Register {
+        id: DataSinkId,
+        data_sink: Box<dyn DataSink>,
+    },
+    Deregister {
+        id: DataSinkId,
+    },
+}
+
+pub struct DataSinkHandle {
```
[the remark this answers is not in the document]
Comment (zhongzc) [reply components/resource_metering/src/reporter/data_sink_reg.rs:58]: Okay

**Review by breeswish — APPROVED** [review]

File: components/resource_metering/src/reporter/data_sink.rs [hunk 5]
```diff
@@ -8,7 +8,7 @@ use kvproto::resource_usage_agent::ResourceUsageRecord;
 
 /// This trait abstracts the interface to communicate with the remote.
 /// We can simply mock this interface to test without RPC.
-pub trait DataSink {
+pub trait DataSink: Send {
     // `try_send` pushes a report data into the sink, which will later be sent to a target
     // by the sink. A deadline can be specified to control how late it should be sent.
     // If the sink is kept full and cannot schedule a send within the specified deadline,
```
Comment (mornyx) [inline components/resource_metering/src/reporter/data_sink.rs:14]: There is no `deadline` in TiKV?

File: components/resource_metering/src/reporter/data_sink_reg.rs [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+use crate::DataSink;
+
+use std::sync::atomic::{AtomicU64, Ordering};
+
+use tikv_util::warn;
+use tikv_util::worker::Scheduler;
+
+/// `DataSinkRegHandle` accepts registrations of [DataSink].
+///
+/// [Collector]: crate::collector::Collector
```
Comment (mornyx) [inline components/resource_metering/src/reporter/data_sink_reg.rs:13]:
```suggestion
/// `DataSinkRegHandle` accepts registrations of [DataSink].
///
/// [DataSink]: crate::DataSink
```

File: components/resource_metering/src/reporter/mod.rs [hunk 7]
```diff
@@ -138,18 +153,57 @@ impl Display for Task {
...
+    }
+
+    pub fn notify(&self, config: Config) {
+        if let Err(err) = self.scheduler.schedule(Task::ConfigChange(config)) {
+            warn!("failed to schedule Task::ConfigChange"; "err" => ?err);
+        }
     }
 }
 
+/// Constructs a default [Recorder], start it and return the corresponding [Scheduler], [DataSinkRegHandle] and [LazyWorker].
```
Comment (mornyx) [inline components/resource_metering/src/reporter/mod.rs:181]:
```suggestion
/// Constructs a default [Recorder], start it and return the corresponding [ConfigChangeNotifier], [DataSinkRegHandle] and [LazyWorker].
```

**Comment by mornyx** [comment]
LGTM

**Review by andylokandy — APPROVED** [review]

**Pull request merged**